### DISCUSSION TITLE
zsock_select() timeout value implementation

### DISCUSSION THREAD

**Discussion by bob2oneil** [zephyrproject-rtos/zephyr#36072] [Q&A] [answered]
I would like to use the equivalent of the Linux **select**() statement, which I prefer over the **poll**() multiplexed i/o in part because of the timeout value being specified in microseconds, whereas **poll**() is in milliseconds. So I would prefer the higher precision in timing so I can use the timeout to yield but have high precision towards timing based events.

Looking through the source code for **zsock_select**(), it appears that underneath the covers it actually uses **zsock_poll**() in the source file **sockets_select.c**, and as a result, it simply converts the microsecond precision timeval into milliseconds.

Do I have this understanding correct in terms of **zsock_select** actual implementation?  Is they any other form of single thread multiplexed i/o based on BSD sockets that might provide higher timer fidelity?

Assuming my observation is correct, does anyone have an opinion on why the poll() implementation is considered superior?

Here is the relevant code from this source file.

```
        poll_timeout = -1;
	if (timeout != NULL) {
		poll_timeout = timeout->tv_sec * 1000
			       + timeout->tv_usec / 1000;
	}

	res = zsock_poll(pfds, num_pfds, poll_timeout);
	if (res == -1) {
		return -1;
	}
```

**Comment by jukkar** [comment, marked as answer]
Indeed, select() is using poll() atm. After looking the code, I think it is trivial to change the code to internally use higher precision in poll implementation and just convert to desired precision in select/poll when calling the internal poll implementation. Patches are welcome if you want to go this route.
We also support `SO_RCVTIMEO` that allows application to define the receive timeout in microsecond resolution.

**Reply by bob2oneil** [reply]
Thanks Jukka,

If I felt comfortable enough in Zephyr to offer up a patch, I would do so.   I could try and code something up, but that would have to be reviewed by the broader team with domain knowledge for unintended consequences.

Your implementation is consistent with the Linux poll() specification at milliseconds resolution, but of course, the wrapper for select() is not.  

If there was a patch, seeing as it appears that Zephyr is somewhat frequently updated, how would that type of change make it into the formal distribution?  Would I just offer up a description of an enhancement for select() compatibility?

**Reply by bob2oneil** [reply]
I may want to pursue this a bit in the future, that is getting a more close fidelity to the Linux select().  You mentioned as an expert that it would not be that difficult for YOU to do some changes.  Me as a newbie to Zephyr would not be as accomplished.  So how would a change request list this get initiated, and would there be any reason for you fine folks to do it considering the pile of other stuff you have to do?  If you could not get to it, I could potentially take a whack at it, but would benefit from some general and specific source files or an broad outline on how you would attack the issue.

**Reply by jukkar** [reply]
What I meant in my earlier comment,  is that we could:

* Separate most of the poll code in `subsys/net/lib/sockets/sockets.c` function `z_impl_zsock_poll()` to `zsock_poll_internal()` and make it to have timeout as a `k_timeout_t` value, then just call the new internal function from `z_impl_zsock_poll()` with msec resolution
* In the `select()` implementation, instead of calling `zsock_poll()` directly, call  `zsock_poll_internal()` with microsecond resolution

This would then allow select() to use microsecond resolution and still use the poll code and not have to create a new implementation to it.

**Reply by bob2oneil** [reply]
Makes sense, I see where you are coming from.  Is it reasonable to think I could create a replacement for the zsock_select() that uses the logic you suggest here, or is this a change that would need to be made to the broader framework such as the creation of the **zsock_poll_internal()** method, and then presumably make it into a formal release at some date?  Would this be something that I should initiate in terms of a useful feature to get it into the queue of future changes?

**Reply by jukkar** [reply]
>  Is it reasonable to think I could create a replacement for the zsock_select() that uses the logic you suggest here,

I would just suggest that you send a PR that proposes this change. The changes look quite simple and they make sense so that select() could be made to work with microsecond resolution. We can then review and tune the PR if there is need to do that.

**Reply by bob2oneil** [reply]
Thanks Jukka, just issued a Pull Request, will see how it goes.  Is there an expectation that I will provide the changes, or is this just a wish list?

**Reply by jukkar** [reply]
>  just issued a Pull Request, will see how it goes. Is there an expectation that I will provide the changes, or is this just a wish list?

~A pull request should contain the changes you are proposing, I am not seeing any from you in https://github.com/zephyrproject-rtos/zephyr/pulls~
Ah, it was this one #36145

**Reply by bob2oneil** [reply]
My bad, I did not fully understand using the PR as the request mechanism vs submitted changed code.  I understood that there was a chance of changed code, but that it might be something one of the reviewers picked up.  I can give it a chance myself, but lacking your domain knowledge, I may not do as good a job.  I also perhaps did not appreciate perhaps how much this is a community development with community submittals.  Thanks for continuing my education into all things Zephyr, as newbie, I am just finding my way.

Can you give this one a look?  Looking through the net_if code for a promiscuous socket, there is packet cloning and it would seem that maybe that is not compatible with a different raw socket.  I need support (meaning stack has to ARP respond) for 257 IP addresses with a UDP only unicast primary IP 192.168.0.1, and then for the 200.168.0.0/24 network more like a raw socket where I need the IP/UDP/TCP header intact, but I can probably recreate the 12 byte Ethernet prior to sending to the destination.

It is not clear to me how I combine the kernel-based net_if stuff with 2 bridge sockets.

I am sure you know the answer.

And once I have the answer, I will likely not need additional Zephyr support for a bit as the squeeking wheel will have its grease.

https://github.com/zephyrproject-rtos/zephyr/discussions/36151

**Reply by jukkar** [reply]
Lets not mix the things in #36151 to this smallish change for select timeout value, those are quite orthogonal. I try to find out some time to implement the change to select() so we get this one forward.

**Comment by jukkar** [comment]
I have a fix proposal at #36220

**Reply by jukkar** [reply]
Actually this is not a discussion but a real issue, unfortunately I do not see a way to convert to an issue.

**Reply by bob2oneil** [reply]
Thanks Jukka for spearheading this.

**Comment by pfalcon** [comment]
In my review comment https://github.com/zephyrproject-rtos/zephyr/pull/36220#pullrequestreview-683875338 , I explain why select() in Zephyr was implemented the way it is, and those were deliberate requirements. Theoretical higher resolution of select() is just that - theoretical benefit, which in practice would only lead to confusion - 1 microsecond passed to select() would likely get rounded up to much higher value (if not on the order of milliseconds, then tenth of milliseconds).

As I mention there, if not thinking in terms of synthetic resolutions, but in algorithmic complexity/conversion overheads, the only way forward is to implement epoll(), but that takes timeout in milliseconds either.

There're a lot of materials on the topic by just googling "select vs poll", here's random google hit: https://devarea.com/linux-io-multiplexing-select-vs-poll-vs-epoll/ and quote:

> Unlike select(), with its inefficient three bitmask-based sets of file descriptors ...

**Reply by bob2oneil** [reply]
Some good information.  But borrowing from my experience with embedded Linux, using select() as a timeout mechanism, I was able to get microsecond precision timing (give or take a few hundred microseconds) based on the timeout value provided.  But that was embedded Linux, and this is perhaps a different animal.

**Comment by bob2oneil** [comment]
Thanks Jukka, for spending the time and getting this into the main branch.  It would not have happened without you handling the logistics and leveraging both you knowledge of the Zephyr network stack as well as the primary contributors to the project necessary for reviewing the effort and getting it into the main repo.
